### PR TITLE
Consolidate ToneAlarm_Linux, ToneAlarm_PX4, and ToneAlarm_ChibiOS

### PR DESCRIPTION
**Pull request by jschall**
This is a work in progress. I'd like to unify all of the AP_Notify_ToneAlarm_XXXXXX files into one AP_Notify_ToneAlarm based on AP_Notify_ToneAlarm_PX4.

Edit: Everything is ready except for AP_HAL_Linux.

### CODE REVIEW

**Comment by tridge** [comment]
This would only make sense if we also converted HAL_Linux. The idea of using RTTTL was so we ended up with a single format after dropping HAL_PX4.
What is the advantage of converting to MML?

**Comment by jschall** [comment]
> What is the advantage of converting to MML?
I personally like MML over RTTTL because I am more familiar with it, I've written a javascript tool to play them in a browser (https://jschall.github.io/), it is more compact, and there are loads of tones available on the web (e.g. https://archeagemmllibrary.com/), I could easily find an online conversion tool for RTTTL->MML (http://www.netmftoolbox.com/tools/rtttl_to_mml.php - haven't tested it), but not the reverse. I like AP_Notify/ToneAlarm_PX4 because it is by far the most refined AP_Notify ToneAlarm backend. It happens to be based on MML.

It is also worth saying, I started on this because I want Matternet's pilots and technicians to experience continuity in behavior between the ChibiOS and NuttX versions of ardupilot.

> This would only make sense if we also converted HAL_Linux. The idea of using RTTTL was so we ended up with a single format after dropping HAL_PX4.
Totally agree, and to be clear, the intention of this WIP PR is to make MML the sole format.

**Comment by jschall** [comment]
I'd like to make the MMLPlayer a member of AP_Notify::ToneAlarm and change AP_HAL::Util to have just a toneAlarm_set_buzzer_tone(frequency,volume) function, but that means that AP_Notify::ToneAlarm will need a 1khz callback from the tonealarm thread in order to update the MMLPlayer and set the buzzer state. How would I hook that up?

**Comment by tridge** [comment]
yes, if we can move the MML player into AP_Notify and simplify the HAL that would be great
we just need to make sure it works on HAL_Linux

**Comment by tridge** [comment]
also note this PR https://github.com/ArduPilot/ardupilot/pull/8952
which makes it easier to test tune playing. It is supported by MAVProxy "playtune" command

**Comment by tridge** [comment]
nice work!
we just need HAL_Linux now
(and a rebase)

**Comment by jschall** [comment]
PX4 support added (slightly hacky). Linux support still pending. Any volunteers to do Linux support?

**Comment by auturgy** [comment]
@lucasdemarchi is this something you might have time to look at?

File: libraries/AP_Notify/ToneAlarm.cpp [hunk 1]
```diff
@@ -392,6 +393,20 @@ void ToneAlarm::handle_play_tune(mavlink_message_t *msg)
 
     mavlink_msg_play_tune_decode(msg, &packet);
 
+    if (_sem) {
+        _sem->take(HAL_SEMAPHORE_BLOCK_FOREVER);
+        _mml_player.stop();
+
+        strncpy(_tone_buf, packet.tune, MIN(sizeof(packet.tune), sizeof(_tone_buf)-1));
+        _tone_buf[sizeof(_tone_buf)-1] = 0;
```
Comment (lucasdemarchi) [inline libraries/AP_Notify/ToneAlarm.cpp]: this is wrong if sizeof(packet.tune) is less than sizeof(_tone_buf)

**Comment by lucasdemarchi** [comment]
@auturgy I won't have much time during next weeks to take a look.

@jschall any particular reason the linux side was not implemented? It shouldn't be so different from the changes needed for the other platforms.

**Comment by auturgy** [comment]
No worries - just thought I’d ask :)

Regards,

James

File: libraries/AP_Notify/ToneAlarm.cpp [hunk 2]
```diff
@@ -337,44 +351,54 @@ void ToneAlarm_PX4::update()
...
+        uint8_t len = strlen(_tone_buf);
+        uint8_t len2 = strnlen(packet.tune2, sizeof(packet.tune2));
+        len2 = MIN((sizeof(_tone_buf)-1)-len, len2);
+        strncpy(_tone_buf+len, packet.tune2, len2);
+        _tone_buf[sizeof(_tone_buf)-1] = 0;
+        _sem->give();
+    }
+
     play_string(packet.tune);
```
Comment (tridge) [inline libraries/AP_Notify/ToneAlarm.cpp]: why is it still calling play_string?

File: libraries/AP_HAL_Linux/ToneAlarm_Disco.cpp [hunk 3]
```diff
@@ -25,58 +25,22 @@ extern const AP_HAL::HAL &hal;
...
-{
-    uint32_t cur_time = AP_HAL::millis();
-
-    if (tune_num != prev_tune_num){
-        tune_changed = true;
-        return true;
+    if (is_zero(frequency) || is_zero(volume)) {
+        bebop_out->play_note(0, 0, 0);
+    } else {
+        bebop_out->play_note(TONEALARM_PWM_POWER, (uint16_t)roundf(frequency), 10000);
```
Comment (tridge) [inline libraries/AP_HAL_Linux/ToneAlarm_Disco.cpp]: I've tested this and unfortunately it doesn't work - it does need to know the duration. I've modified it to add the correct duration and it works nicely on Disco

**Comment by tridge** [comment]
rebased and fixed an issue with starting tone_alarm on HAL_PX4

**Pull request merged**